media/simulateSkill/simulateSkill.js

```diff
@@ -412,6 +424,13 @@ async function handleMessageFromExtension(event) {
             setHTMLElementStatuses(ENABLE);
         }
     }
+    else if (message.type === SIMULATOR_MESSAGE_TYPE.UPDATE_APL_VIEW) {
+        const previousState = vscode.getState();
+        const tabToShow = previousState.currentTabGlobal;
+        if (tabToShow === APL_VIEW_TAB && previousState.aplViewport !== undefined && previousState.aplDocument !== undefined && previousState.aplDocument !== '') {
+            await updateAplViewPort(previousState.aplDocument, previousState.aplDatasource, previousState.aplViewport);
```

Comment:
This seems alright. However, since you are setting the state and hence re-rendering the APL on simulate view, if the APL template has an audio or a video tag, it seems to replay it again, whenever the user switches the webview. I don't think that is a good experience. Also, I don't understand what you mean by the whole APL not being rendered when you switch tabs. Can you please provide more detail?

Reply:
yeah, you are right that rerender will cause replay foe some APL template. The whole APL not being rendered means sometimes the preview is not complete, like below, should have `vegetable` under fruit. Maybe has some storage limitation on VSC webview or VSC processing something when switch webview, this is also one reason why they put APL preview in a single view.
<img width="1015" alt="Screen Shot 2021-01-21 at 11 57 08" src="https://user-images.githubusercontent.com/68405785/106831362-a19a8600-6644-11eb-85bc-aa820aac90ca.png">
I guess we need to do option between these two behaviors(not totally rendered / replay some element) when user reopen webview, do you think which one better for users?

Reply:
As you can see under the persistence section of [webview api docs](https://code.visualstudio.com/api/extension-guides/webview#persistence), we use the basic `getState` and `setState` methods, to persist the view . As per the docs, this is efficient. However, this could be leading to the issue you mentioned above. 

We can see if changing the persistence method can help us here. However, I wonder how widespread this issue is. Typically, the expectation would be, skill developers would simulate the whole skill session, before moving out of the simulator view. In such cases, it shouldn't be a huge issue on what the apl response rendering would be, since the session is closed.

Do you think this is not a good assumption? Would like to hear your thoughts on this.

Reply:
I agree that most of the time, user will not change webview. The scenario I concerned is testing with local debugging. When uses local debugging tool, the breakpoints will switch the view to code, if users come back to simulator, they can't continue the test if the APL preview is not complete.

Reply:
And for the issue that replay after recovering webview, I found even without rerender APL, some element will replay too. Because we store the html page and reload the html page after switch webviews, lol.
Like the video below, the test disappear when user click on the button(this command is already in the current release as it's integrated in APL renderer. This command is different with click event that will not send a user_event, just affect the preview.), and will recover after user switch the webview. So do you have other concerns on this PR? :) 

https://user-images.githubusercontent.com/68405785/106939828-a22f2d00-66d5-11eb-8f0c-04ce916d892e.mov

Reply:
As discussed offline, lets' see if we can find a different persistence approach for the simulator webview instead of re-rendering it.